packages/tpetra/core/src/Tpetra_Distributor.cpp

```diff
@@ -554,9 +566,32 @@ namespace Tpetra {
       std::ostringstream os;
       os << myRank << ": doWaits: # reqs = "
          << requests_.size () << endl;
+      os << myRank << ": doWaits: # persistentReqs = "
+         << persistentRequests_.size () << endl;
       *out_ << os.str ();
     }
 
+    int count = persistentRequests_.size();
```

Comment:
1. Use of raw MPI needs to be protected by the `HAVE_TPETRACORE_MPI` macro.
2. Just call `Teuchos::waitAll`; it does exactly what lines 578-586 do.
3. We should talk about better encapsulations (e.g., a "CommRequests" interface with an `MpiCommRequests` base class that has an array of raw `MPI_Request` in it).

Reply:
As far as I can see, 'waitAll' actually frees the requests. I can add an optional parameter to take care of that. And add a 'startAll' function.